src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,249 @@
-end)
+    In the basic derivation {[ 'a list ]} would be translated to {[gen_list]}. However,
+    we want the generator to call itsef.
+ *)
 
 let rec longident_to_str = function
   | Lident s -> s
   | Ldot (lg, s) -> Printf.sprintf "%s.%s" (longident_to_str lg) s
   | Lapply (lg1, lg2) ->
       Printf.sprintf "%s %s" (longident_to_str lg1) (longident_to_str lg2)
 
-let rec is_rec_typ typ_name = function
+(** [is_rec_type_decl typ_names typ] looks for elements of [typ_names]
+    recursively in [typ]. *)
+let rec is_rec_type_decl typ_names typ =
+  let in_type_kind =
+    match typ.ptype_kind with
+    | Ptype_variant cstrs -> List.exists (is_rec_constr_decl typ_names) cstrs
+    | _ -> false
+  in
+  let in_manifest =
+    match typ.ptype_manifest with
+    | Some x -> is_rec_typ typ_names x
+    | None -> false
+  in
+  in_type_kind || in_manifest
+
+and is_rec_typ typ_names = function
   | { ptyp_desc = Ptyp_constr ({ txt = x; _ }, _); _ } ->
-      longident_to_str x = typ_name
-  | { ptyp_desc = Ptyp_tuple xs; _ } -> List.exists (is_rec_typ typ_name) xs
+     List.exists (fun typ_name -> longident_to_str x = typ_name) typ_names
+  | { ptyp_desc = Ptyp_tuple xs; _ } -> List.exists (is_rec_typ typ_names) xs
+  | { ptyp_desc = Ptyp_variant (rws, _, _); _ } ->
+     List.exists (fun rw ->
+         match rw.prf_desc with
+         | Rtag (lab, _, cts) ->
+            List.exists (fun typ_name -> lab.txt = typ_name) typ_names ||
+              List.exists (is_rec_typ typ_names) cts
+         | Rinherit ct -> is_rec_typ typ_names ct) rws
   | _ -> false
 
-let name s =
-  let prefix = "gen" in
-  match s with "t" -> prefix | s -> prefix ^ "_" ^ s
+and is_rec_constr_decl typ_names cd =
+  match cd.pcd_args with
+  | Pcstr_tuple cts -> List.exists (is_rec_typ typ_names) cts
+  | _ -> false
 
-let pat ~loc s =
-  let (module A) = Ast_builder.make loc in
-  let s = name s in
-  A.pvar s
+(** [env] contains the list of recursive types during the derivation
+
+    i.e:
+    - contains one element maximum if its a single type declaration *)
+type env = string list ref
+
+(** [curr_types] saves both the current type and all mutual
+    recursive types *)
+let curr_types : (string * string list) option ref = ref None
+
+let set_curr_types x xs = curr_types := Some (x, xs)
+
+let clean_curr_types () = curr_types := None
 
-let gen ~loc ?(env = TypeGen.empty) lg =
+let curr_type () = Option.get !curr_types |> fst
+
+let curr_types () = Option.get !curr_types |> snd
+
+let env : env = ref []
+
+let clean_env () = env := []
+
+let is_rec x = List.mem x !env
+
+let add_env typ_names typ_name ty =
+  if is_rec_type_decl typ_names ty then
+    env := typ_name :: !env
+
+
+(** {2. Generator constructors} *)
+
+(** [gen lg args] creates a generator using [lg].
+
+    The longident can either be a:
+    - Lident s: We transform to gen_s (or gen if s = "t")
+    - Ldot (lg, s): We transform to qualified generator (e.g. B.gen)
+*)
+let gen ~loc lg args =
```

Comment:
`gen` is not very descriptive (it's all generators).
`gen_ident` or `gen_longident` perhaps?

Reply:
You are right: d527aa353bf3f971791ff1a3e112d8d2d24179d1